src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -0,0 +1,232 @@
+            // sends additional VS specific capabilities, so directly deserialize them into the VSClientCapabilities
+            // to avoid losing them.
+            this._clientCapabilities = input["capabilities"].ToObject<VSClientCapabilities>();
+
+            return new InitializeResult
+            {
+                Capabilities = new VSServerCapabilities
+                {
+                    DocumentHighlightProvider = true,
+                    DocumentSymbolProvider = true,
+                    DocumentFormattingProvider = true,
+                    DocumentRangeFormattingProvider = true,
+                    DocumentOnTypeFormattingProvider = new DocumentOnTypeFormattingOptions { FirstTriggerCharacter = "}", MoreTriggerCharacter = new[] { ";", "\n" } },
+                    DefinitionProvider = true,
+                    CompletionProvider = new CompletionOptions { ResolveProvider = true, TriggerCharacters = new[] { "." } },
+                    ImplementationProvider = true,
+                    ReferencesProvider = true,
+                    SignatureHelpProvider = new SignatureHelpOptions { TriggerCharacters = new[] { "(", "," } },
+                    WorkspaceSymbolProvider = true,
+                }
+            };
+        }
+
+        [JsonRpcMethod(Methods.InitializedName)]
+        public Task Initialized() => Task.CompletedTask;
+
+        [JsonRpcMethod(Methods.ShutdownName)]
+        public object? Shutdown(CancellationToken _) => null;
+
+        [JsonRpcMethod(Methods.ExitName)]
+        public void Exit()
+        {
+        }
+
+        /// <summary>
+        /// Go to definition calls to get third party definitions (xaml) and to get those definitions it expectes to be called on the UI thread.
+        /// <see cref="VisualStudioSymbolNavigationService.WouldNotifyToSpecificSymbol"/>
+        /// </summary>
+        [JsonRpcMethod(Methods.TextDocumentDefinitionName)]
+        public async Task<object> GetTextDocumentDefinitionAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var textDocumentPositionParams = input.ToObject<TextDocumentPositionParams>();
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            return await this._protocol.GoToDefinitionAsync(_workspace.CurrentSolution, textDocumentPositionParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentCompletionName)]
+        public async Task<object> GetTextDocumentCompletionAsync(JToken input, CancellationToken cancellationToken)
+        {
+            var completionParams = input.ToObject<CompletionParams>();
+            return await this._protocol.GetCompletionsAsync(_workspace.CurrentSolution, completionParams, _clientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
+
+        [JsonRpcMethod(Methods.TextDocumentCompletionResolveName)]
```

Comment:
It still seems odd to me we more or less get a request in with a JSON payload and a method name. We unpack this and call an method on our LanguageServerProtocol, which then goes and looks up the handler *back* by the string again...feels some some abstraction here isn't right.

Reply:
True, however this way makes it easier to share the code between this and the exports for liveshare, I can easily just lookup the handler by LSP method.  Once I can delete the rest of the liveshare exports, then I can revisit, for now I won't change in this PR as it's gotten large.

It may also be possible to 'pass through' the json payload directly to the handler with streamjsonrpc.  I did look into it a bit, but handling it with the generic parameters and return types didn't seem straightforward.